R/ComputeFirstDifferences.R

```diff
@@ -0,0 +1,53 @@
+#' Adds first differences of a variable, robustly to missing values, as new columns in a panel dataset
+#'
+#' @param df Data frame that will be modified
+#' @param idvar Character indicating column of grouping variable.
+#' @param timevar Character indicating column of time variable.
+#' @param diffvar Character indicating column of variable whose first difference will be taken.
+#' @param timevar_holes Logical indicating whether the panel contains missing values in the time variable.
```

Comment:
Do we literally mean "missing values" or do we mean something more like "gaps"?

Reply:
That's a great point @jmshapir. If there are literal NAs in the data, but all time periods are present then there wouldn't be a problem. The issue arises when there are implicit NAs that do not appear in the data. 

"gaps" seems like a more appropriate word, so I changed wording in https://github.com/JMSLab/eventstudyr/pull/38/commits/12d93671d0f44992cf1ad63e265a3810ccf77431. I also implemented other improvements to the description of these new functions.

Reply:
@santiagohermo thanks! Per my question in https://github.com/JMSLab/eventstudyr/pull/38#pullrequestreview-1419618439, can you let me know what is the intended behavior here when the time value has gaps? I want to make sure we don't inadvertently create side effects for users. Thanks!

Reply:
The intended default behavior is to impute NAs to any observation with any value in the diff or lag missing. For instance:

| i | t | z | z_diff |
| --- | --- | --- | --- |
| 1 | 1 | 0 | 0 |
| 1 | 2 | 0 | 0 |
| 1 | 4 | 1 |  NA |
| 1 | 5 | 1 |  0 |
| 2 | 1 | 0 | 0 |
| 2 | 2 | 0 | 0 |
| 2 | 3 | 1 | 1 |
| 2 | 5 | 1 |  NA |
 
With this behavior some rows will end up with NAs in the event-study variables, and thus will be automatically dropped in estimation.

We also issue a message saying that "dataset contains holes in time variable", see [here](https://github.com/JMSLab/eventstudyr/blob/d4fd6859a4763bbac34d1286752a10b8553bc50b/R/EventStudy.R#L191).

Reply:
FYI @jmshapir 

One more comment. Because I thought we were forcing the time column to be an integer, I created the hole detection algorithm to check whether the difference in the (sorted) time variable is different than 1 ([here](https://github.com/JMSLab/eventstudyr/blob/d4fd6859a4763bbac34d1286752a10b8553bc50b/R/EventStudy.R#L184)). However, we ask the user to use a numeric column as time var, which need not be an integer ([here](https://github.com/JMSLab/eventstudyr/blob/d4fd6859a4763bbac34d1286752a10b8553bc50b/R/EventStudy.R#L165)). So  in https://github.com/JMSLab/eventstudyr/pull/38/commits/64d281bfb65b54426f9844e09c712fc9e31ae260 I added a check to make sure that the timevar column only contains integeres.

We could allow for arbitrary types of time variable by either:
1. Asking the user to tell us the "step" of the time variable.
2. Creating a masked time variable that is just an increasing count of integers and use it inside the package.

This sounds like a task for a new issue though.

Reply:
@santiagohermo thanks!

On the [intended default behavior](https://github.com/JMSLab/eventstudyr/pull/38#discussion_r1194120884) that sounds good (and pretty safe), thanks!

On [non-integer time values](https://github.com/JMSLab/eventstudyr/pull/38#discussion_r1194142080), I think a good solution might be to enable (2) as an option for the user. (So, at the user's request, we can transform any time variable into an integer one.) But, since this is pretty easy for the user to do themselves, I don't think it's pressing. How about if you add it to the [someday](https://github.com/JMSLab/eventstudyr/wiki/Roadmap/ddb43b079f99d19a109d2781bfbda4fabe4f410b#someday) section of the roadmap so we'll remember the discussion?

After that you can resolve this thread thanks!

Reply:
Thanks @jmshapir! I added the point on non-integer time values to the Roadmap, see [here](https://github.com/JMSLab/eventstudyr/wiki/Roadmap/_compare/ddb43b079f99d19a109d2781bfbda4fabe4f410b...55daa41ac092e44afc67af51dd929291f69bd131).

Reply:
@santiagohermo thanks for adding the [table](https://github.com/JMSLab/eventstudyr/pull/38#discussion_r1194120884), it's really helpful!

Confirming that we want _zdiff_ at $t = 1$ to equal 0 (rather than NA) for all i?

Reply:
Thanks @nateschor! Assuming that t runs from 1 to 5, the example is wrong and the behavior you describe is actually the default. I was focusing on how the function behaves with gaps so I was not paying attention to endpoints, apologies for the mistake!

Updated example:

| i | t | z | z_diff |
| --- | --- | --- | --- |
| 1 | 1 | 0 | NA |
| 1 | 2 | 0 | 0 |
| 1 | 4 | 1 |  NA |
| 1 | 5 | 1 |  0 |
| 2 | 1 | 0 | NA |
| 2 | 2 | 0 | 0 |
| 2 | 3 | 1 | 1 |
| 2 | 5 | 1 |  NA |